akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolGateway.scala

```diff
@@ -4,91 +4,89 @@
-    val ref = system.actorOf(props, PoolInterfaceActor.name.next())
-    new AtomicReference[State](Running(ref, _shutdownStartedPromise, shutdownCompletedPromise))
+    gatewayRef ! Shutdown(this, shutdownCompletedPromise)
+    shutdownCompletedPromise.future
+  }
+
+  override def toString = s"PoolGateway(hcps = $hcps)"
+
+  // INTERNAL API (testing only)
+  private[client] def poolStatus(): Future[Option[PoolInterfaceStatus]] = {
+    val statusPromise = Promise[Option[PoolInterfaceStatus]]()
+    gatewayRef ! PoolStatus(this, statusPromise)
+    statusPromise.future
   }
 
-  def currentState: Any = state.get() // enables test access
-
-  def apply(request: HttpRequest, previousIncarnation: PoolGateway = null): Future[HttpResponse] =
-    state.get match {
-      case Running(ref, _, _) ⇒
-        val responsePromise = Promise[HttpResponse]()
-        ref ! PoolInterfaceActor.PoolRequest(request, responsePromise)
-        responsePromise.future
-
-      case IsShutdown(shutdownCompleted) ⇒
-        // delay starting the next pool incarnation until the current pool has completed its shutdown
-        shutdownCompleted.flatMap { _ ⇒
-          val newGatewayFuture = Http().cachedGateway(hcps)
-          // a simple set is fine here as `newGatewayFuture` will be identical for all threads getting here
-          state.set(NewIncarnation(newGatewayFuture))
-          apply(request)
-        }
-
-      case x @ NewIncarnation(newGatewayFuture) ⇒
-        if (previousIncarnation != null)
-          previousIncarnation.state.set(x) // collapse incarnation chain
-        newGatewayFuture.flatMap(_(request, this))
+  override def equals(that: Any): Boolean =
+    that match {
+      case p: PoolGateway ⇒ p.hcps == hcps && p.gatewayId == gatewayId
+      case _              ⇒ false
     }
 
-  // triggers a shutdown of the current pool, even if it is already a later incarnation
-  @tailrec final def shutdown(): Future[Done] =
-    state.get match {
-      case x @ Running(ref, shutdownStartedPromise, shutdownCompletedPromise) ⇒
-        if (state.compareAndSet(x, IsShutdown(shutdownCompletedPromise.future))) {
-          shutdownStartedPromise.success(Done) // trigger cache removal
-          ref ! PoolInterfaceActor.Shutdown
-          shutdownCompletedPromise.future
-        } else shutdown() // CAS loop (not a spinlock)
+  override def hashCode(): Int = hcps.hashCode() ^ gatewayId.hashCode()
+}
+
+private[http] object PoolGateway {
 
-      case IsShutdown(x)                    ⇒ x
+  sealed trait GatewayIdentifier
+  case object SharedGateway extends GatewayIdentifier
+  final case class UniqueGateway(id: Long) extends GatewayIdentifier
 
-      case NewIncarnation(newGatewayFuture) ⇒ newGatewayFuture.flatMap(_.shutdownAux())
-    }
+  private[this] val uniqueGatewayId = new AtomicLong(0)
+  def newUniqueGatewayIdentifier = UniqueGateway(uniqueGatewayId.incrementAndGet())
```

Comment:
Since the gateways are all children of the same parent actor, I don't think this AtomicLong is needed anymore.
A single var in the parent actor would be sufficient